conftest.py

```diff
@@ -760,3 +786,257 @@ def cifs_iso_sr(host, cifs_iso_device_config):
+                test_vms.add(fixture_value)
+                test_hosts.add(fixture_value.host)
+            # Check if fixture is a list
+            elif isinstance(fixture_value, list):
+                for item in fixture_value:
+                    if isinstance(item, Host):
+                        test_hosts.add(item)
+                    elif isinstance(item, VM):
+                        test_vms.add(item)
+                        test_hosts.add(item.host)
+            # Check if fixture is a dict with Host or VM values
+            elif isinstance(fixture_value, dict):
+                for value in fixture_value.values():
+                    if isinstance(value, Host):
+                        test_hosts.add(value)
+                    elif isinstance(value, VM):
+                        test_vms.add(value)
+                        test_hosts.add(value.host)
+            # Check if fixture has a 'host' attribute (like SR, Pool objects)
+            elif hasattr(fixture_value, 'host') and isinstance(fixture_value.host, Host):
+                test_hosts.add(fixture_value.host)
+            # Check if fixture is a Pool
+            elif hasattr(fixture_value, 'hosts') and isinstance(fixture_value.hosts, list):
+                test_hosts.update(h for h in fixture_value.hosts if isinstance(h, Host))
+        except Exception:
+            # Some fixtures may not be available yet or may fail to load
+            pass
+
+    yield
+
+    # After test completes, check if it failed
+    report = request.node.stash.get(PHASE_REPORT_KEY, None)
+    if report and "call" in report and report["call"].failed:
+        test_nodeid = request.node.nodeid
+        logging.debug(f"Test failed: {test_nodeid}, "
+                      f"used hosts: {[h.hostname_or_ip for h in test_hosts]}, "
+                      f"VMs: {[vm.uuid for vm in test_vms]}")
+
+        if test_hosts:
+            FAILED_TESTS_HOSTS[test_nodeid] = test_hosts
+            SESSION_HAS_FAILURES = True
+        if test_vms:
+            FAILED_TESTS_VMS[test_nodeid] = test_vms
+
+            # Check if test/class has capture_console marker for console screenshots
+            has_console_marker = request.node.get_closest_marker("capture_console") is not None
+            if has_console_marker:
+                logging.debug(f"Capturing console NOW...")
+
+                # Capture console for each VM immediately (while VM is still alive)
+                # Use the session-wide log directory
+                for vm in test_vms:
+                    try:
+                        screenshot_path = capture_vm_console(vm, session_log_dir)
+                        logging.debug(f"Console captured: {screenshot_path}")
+                    except Exception as e:
+                        logging.error(f"Failed to capture console for VM {vm.uuid}: {e}")
+
+@pytest.fixture(scope='session', autouse=True)
+def collect_logs_on_session_failure(request, hosts, session_log_dir):
```

Comment:
IMO this function is quite chatty logging-wise.

Reply:
You're right. I tested in debug and didn't noticed that. I removed some put some others in debug.